src/forms/form.rs

```diff
@@ -0,0 +1,355 @@
+    pub fn get_field<T>(&mut self, name: &str) -> Option<Box<T>>
+    where
+        T: FormField + 'static,
+    {
+        for (id, (field_name, _)) in self.fields.iter().enumerate() {
+            if name == *field_name {
+                let (_, field) = self.fields.remove(id);
+                return field.to_any().downcast::<T>().ok();
+            }
+        }
+        None
+    }
+
+    /// Get the output of a specific field if it exists within the Form
+    pub fn get_result(&self, name: &str) -> Option<FormValue> {
+        for (field_name, field) in self.fields.iter() {
+            if name == *field_name {
+                return Some(field.get_output());
+            }
+        }
+        None
+    }
+
+    /// Get the errors generated from a specific field.
+    ///
+    /// You must run [is_valid](#methods.is_valid) first in order to be able to retrieve the errors
+    pub fn get_error(&self, name: &str) -> Option<&FormValidationResult> {
+        for (field_name, errors) in self.errors.iter() {
+            if name == *field_name {
+                return Some(errors);
+            }
+        }
+        None
+    }
+
+    /// Change focus on the currently active field
+    fn update_active_field(&mut self) {
+        let mut height = 0;
+        let mut active_min_height = 0;
+        for (id, (_, field)) in self.fields.iter_mut().enumerate() {
+            let active = self.active && id == self.index;
+            field.set_active(active);
+            if active {
+                active_min_height = height;
+            }
+            height += field.get_height();
+            if field.display_label() {
+                height += 1;
+            }
+        }
+        // detect when the active field is outside of the scroll
+        if self.scroll_index < active_min_height as usize
+            || self.scroll_index + self.get_height() as usize > active_min_height as usize
+        {
+            self.scroll_to(active_min_height as usize);
+        }
+    }
+
+    /// Checks whenever the user went through the entire form, and confirmed on the last field
+    pub fn is_finished(&self) -> bool {
+        self.index >= self.fields.len()
+    }
+
+    /// Checks if the form is entirely valid. If any field fails its `validate` method, the function returns `false`
+    /// and a list of every reported error is stored in the form, waiting for [get_error](#methods.get_error) to be called
+    pub fn is_valid(&mut self) -> bool {
+        self.errors.clear();
+        for (name, field) in self.fields.iter() {
+            let mut field_errors: FormValidationResult = vec![];
+            field.validate(&mut field_errors);
+            if !field_errors.is_empty() {
+                self.errors.insert(String::from(name), field_errors);
+            }
+        }
+        let mut self_errors: FormValidationResult = vec![];
+        self.validate(&mut self_errors);
+
+        self.errors.is_empty() && self_errors.is_empty()
+    }
+}
```

Comment:
This function feels weird to me.
We call `validate` on every field and we get their errors, but then we call `self.validate` which also calls `validate` on all the forms and gets the errors.
Also I'd suggest removing this function and just have `self.validate` return the errors and then user will check if errors is empty to see if form is valid. This would make storing errors and `get_errors` obsolete.
Thoughts?

Reply:
I may have been a bit tired when I worked on this, Forms are a special kind of field and I wanted them to be able to retrieve errors by field independently from each other. I'll try to rethink this part properly.